Guide for Getting Started on Gardener with Cloud SDK for Java (#470)
SAP/cloud-sdk: docs/java/environments/sap-btp-kubernetes-environment-with-sap-gardener.mdx

However, this guide will focus on the Service Operator usage.

:::tip
In case you don't have it installed please follow the [documentation](https://github.com/SAP/sap-btp-service-operator#sap-business-technology-platform-sap-btp-service-operator-for-kubernetes) to install it.
:::

</details>

<details>
  <summary>Docker</summary>

This guide assumes you have [Docker](https://www.docker.com/) installed on your local machine.

Furthermore, you need a **Docker repository** where you can store images.
The repository needs to be publicly accessible in order for the cluster to access and download the Docker image we are going to create.

In case you don't have such a repository yet we recommend either:

- [Docker Hub](https://hub.docker.com/)
- [Artifactory DMZ](https://common.repositories.cloud.sap/ui/) (for SAP internal developers)

Access to images in a repository may be limited to authenticated and/or authorized users, depending on your configuration.

Make sure you are logged in to your repository on your local machine by running:

```bash
docker login (your-repo) --username=(your-username)
```

> And check your configuration with `cat ~/.docker/config.json`

Reviewer: As `cat` is specific to linux or mac and I am not sure where the file is stored in a windows machine, consider rewording this to examine the `config.json` file in your docker home directory?

:::tip
In case AuthN/AuthZ is required to download images make sure you have a [secret configured in your cluster](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#create-a-secret-by-providing-credentials-on-the-command-line):

```bash
kubectl create secret docker-registry (name-of-the-secret) --docker-username=(username) --docker-password=(API-token) --docker-server=(your-repo)
```

:::

</details>

<details>
  <summary>Application using the SAP Cloud SDK</summary>

If you don't have an application already you can comfortably [create one from our archetypes](../getting-started).

</details>

### Containerize the Application

To run on Kubernetes the application needs to be shipped in a container.
For this guide we will be using Docker.

Create a `Dockerfile` in the project root directory:

```dockerfile {2}
FROM openjdk:8-jdk-alpine
ARG JAR_FILE=application/target/*.jar
COPY ${JAR_FILE} app.jar
ENTRYPOINT ["java","-jar","/app.jar"]
EXPOSE 8080
```

> If needed update the `JAR_FILE` to point to your JAR file.

Reviewer: ```suggestion
If needed, update the `JAR_FILE` to point to your JAR file.
```

:::tip
You can find more information on how to containerize Spring Boot applications in [this guide](https://spring.io/guides/gs/spring-boot-docker/) (in particular, check the _Containerize It_ section).
:::

Compile and push the image by running:

```bash
docker build -t <your-repo>/<your-image-name> .

> docker push <your-repo>/<your-image-name>

Author: `-t` creates the tag :)

```

:::tip
In case you are facing authorization issues when pushing to your repository refer to the dedicated section under [Prerequisites](#prerequisites).
:::

### Create a Kubernetes Deployment

1. Create a new YAML file describing the Kubernetes deployment:

>    ```yml title="deployment.yml" {18,32}

Author: yep, there are more cool features available even. I found it here: https://docusaurus.io/docs/markdown-features/code-blocks

          - path: /
            pathType: Prefix
            backend:
              service:
                name: my-app
                port:
                  number: 8080
```

2. Install the configuration via `kubectl apply -f ingress.yml`.

3. Verify the Ingress is up and running: `kubectl describe ingress ingress-router`

You should see an entry with the path `/` pointing to the backend `my-app`.

:::note
In case something went wrong and you are struggling to configure the Ingress you can also come back and set it up later.
The Ingress is a convenient way to access your application.
It is not strictly required for the rest of this guide.
:::

<details>
  <summary>Recommended: Configure TLS for your Ingress</summary>
  
Enable the NGINX Ingress add-on in your Gardener dashboard.
The process may take a few minutes.
Afterwards, you should see a domain in the dashboard as well as a Kubernetes secret `secret-tls`.

Un-comment the 4 lines in the YAML above using the generated domain and secret.
Then re-deploy the configuration as usual.
Your cluster endpoint should now be trusted by your browser.

</details>

:::tip
We highly recommended enabling TLS for your cluster endpoints.

> It ensures your client (e.g. browser) can verify the clusters identity.

Reviewer: ```suggestion
It ensures your client (e.g. browser) can verify the cluster's identity.
```

:::

### Access Your Application

At this point take a moment to verify you can access your application.
Use the host you have defined in your Ingress rule in a browser or other tool of your choice (e.g. Postman).
In case you started with an SAP Cloud SDK Archetype your should be greeted with a welcome page under the root path.

:::note
In case you skipped setting up an Ingress before you can use port forwarding to access your application.

Run `kubectl port-forward deployment my-app-deployment :8080`.

> With that you should be able to access the application on your `http://localhost:8080`

Reviewer: ```suggestion
With that you should be able to access the application on your `http://localhost:<port-printed-on-the-console>`
```
I use the port printed out on the console instead of `8080`, have you tested it with `8080`?

Author: I improved it to explicitly give the port that should be forwarded

![Gardener](https://gardener.cloud/images/lp/lp-gardener-diagram.svg)

## SAP Cloud SDK Features Supported on SAP Gardener

> As of version `3.55.0` the SAP Cloud SDK supports SAP Gardener-based Kubernetes clusters out of the box.

Reviewer 2: ```suggestion
As of version `3.55.1` the SAP Cloud SDK supports SAP Gardener-based Kubernetes clusters out of the box.
```

> The following features are covered:

Reviewer 3: ```suggestion
Find below the list of features we currently support:
```

> - ✅ Consume SAP BTP services like Destination, Connectivity, IAS, XSUAA, and others.

Reviewer 3: ```suggestion
- ✅ Consume SAP BTP services like Destination, Connectivity, IAS, XSUAA, and others
```
